Bug 1254567 & Bug 1254568 - Create menu views

fluffyemily opened the pull request:
This is V1 of the menu code. It is structured like this:

`MenuViewController` creates an instance of the `MenuView`, sets display properties of the menu and controls the overall placement and size of the menu on the screen. It implements delegate and datasource protocols for the toolbar and the menu items.

The `MenuView` consists of a toolbar, a menu page controller and a footer. 

The toolbar displays when there are toolbar items available through the toolbar datasource. It displays above the menu pages when being presented modally or below the menu pages when being presented as a popover.

The Menu page controller is a custom page controller, `MenuPagingViewController`. I have done this because `UIPageController` is just broken in so many ways. Attempting to get the menu items to layout in any rational sense when using a `UIPageController` was virtually impossible. In the end I wrote this one in less time that in took to attempt to use AutoLayout within a `UIPageController`. I chose _not_ to create it as a generic all purpose page controller that reuses views and only retains hold of previous/current/next pages as I felt that if we were to actually end up with more than 3 pages in our menu we need to seriously rethink our menu strategy. That does not mean we could not repurpose the page controller for such things later, should we need it.

Each page in the page controller is an instance of `MenuPageViewController`, which uses the menu item datasource to layout and display the appropriate menu items, each menu item being an instance of `MenuItemView`.

The page footer is a simple view that displays only when the menu is being displayed in modal.

Question: Currently, the `MenuView` itself makes these decisions based upon a display type enum - Should this be something that the `MenuViewController` should be able to change instead?

Currently the buttons do nothing, but this will come soon and be controlled via the delegate methods on both toolbar and Menu items.

Things also not yet done - it doesn't handle rotation particularly well right now - I'm working on it.

To display the menu on iPhone and iPad, please add the following snippets somewhere (I used `viewDidLoad` in `BrowserViewController`)

iPhone:
```
        let mvc = MenuViewController(withMenuConfig: MenuConfiguration.menuConfigurationForLocation(.TabTray), presentationStyle: .Modal)
        mvc.modalPresentationStyle = .OverCurrentContext
        self.presentViewController(mvc, animated: false, completion: nil)
```

iPad:
```
        let mvc = MenuViewController(withMenuConfig: MenuConfiguration.menuConfigurationForLocation(.TabTray), presentationStyle: .Popover)
        mvc.modalPresentationStyle = .Popover
        mvc.popoverPresentationController?.sourceView = urlBar
        mvc.popoverPresentationController?.sourceRect = urlBar.frame
        mvc.popoverPresentationController?.permittedArrowDirections = UIPopoverArrowDirection.Up

        self.presentViewController(mvc, animated: false, completion: nil)
```

Also covers bug https://bugzilla.mozilla.org/show_bug.cgi?id=1254568

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+ * creates a view that consists of an image and a title.
+ * the image view is displayed at the center top of the
+ **/
+public class MenuItemView: UIControl {
```
Why subclass `UIControl` instead of `UIButton`? Seems like using UIButton would save you from having to track touches below.

fluffyemily replied:
When I added this in my sample project using just iOS9, adding the custom views to a UIButton subclass caused the button not to display, whereas subclassing the UIControl did work. Just turned the UIControl into a UIButton here and everything layed out just fine, so I guess I was doing something wrong on the sample menu project. Updating.

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+    private var previousLocation: CGPoint?
+    public init() {
+        imageView = UIImageView()
+        titleLabel = UILabel()
```
Nit: Can move these alongside the property definitions above. You can also move the setup code for the label into a lazy block to clean up the init call.

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+        super.init(frame: CGRectZero)
+
+        titleLabel.lineBreakMode = NSLineBreakMode.ByWordWrapping
+        titleLabel.numberOfLines = 2
```
I'm a bit worried constraining this to two lines. It might be better to have it set to 0 lines and have it flow instead. Not sure how this will work with long text locales.

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+            make.centerX.equalTo(self)
+            make.bottom.equalTo(self).offset(-padding)
+            make.width.lessThanOrEqualTo(self.snp_width).dividedBy(1.5)
+        }
```
Would it make it easier to setup the constraints so that:

1. The imageView is pinned to the top by a margin, centered X
2. Pin the top of the label to the bottom of the image and left/right using a margin then pinning the bottom to a greaterThanOrEqual to bottom self with a margin.

If you set the number of lines to 0 and have it flow, the constraints should hold correctly regardless of 1, 2, or 3 lines. Would also clean up the pinning to the width and division by 1.5.

fluffyemily replied:
So, I updated the constraints here to:
```
        customImageView.snp_makeConstraints { make in
            make.centerX.equalTo(self)
            make.top.equalTo(self).offset(2 * padding)
        }

        customTitleLabel.snp_makeConstraints { make in
            make.top.greaterThanOrEqualTo(customImageView.snp_bottom).offset(padding/2)
            make.centerX.equalTo(self)
            make.left.equalTo(self).offset(padding)
            make.right.equalTo(self).offset(-padding)
            make.bottom.greaterThanOrEqualTo(self).offset(-padding)
        }
```

The result of which was a good alignment, however, the single line rows need to be aligned centrally between the image and the bottom of the view, so it appears in the center of the 2 line text rather than at the bottom. Not really sure how to achieve that here (that was kind what all the fancy division was all about). Any ideas?

How it was:
![screen shot 2016-03-21 at 4 46 31 pm](https://cloud.githubusercontent.com/assets/12613048/13926507/8648b3f4-ef84-11e5-93ff-8256e3c1e6f5.png)

Now:

![screen shot 2016-03-21 at 4 44 39 pm](https://cloud.githubusercontent.com/assets/12613048/13926441/4245bb70-ef84-11e5-9666-f1f643383bc6.png)

sleroux replied:
I see. Only think I can think of is putting the label inside a fixed size container and centering the label inside of it. Not sure if that is better than the division math you were doing though. It would look like:

```swift
centerContainer.addSubview(textLabel)
textLabel.snp_makeConstraints { make in
    make.center.equal(centerContainer)
}
```

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+        titleLabel.preferredMaxLayoutWidth = self.bounds.size.width / 1.5
```
Using the revised constraints I think you'll be able to remove this.

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+        imageView.image = nil
+        titleLabel.text = nil
+
+        guard let recognisers = gestureRecognizers else { return }
```
Nit: Should it be recognizers or recognisers?

fluffyemily replied:
I still struggle to spell American sometimes.

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+            }
+        }
+
+        self.snp_removeConstraints()
```
I noticed that above you're setting the views to nil and removing the constraints. This would cause each cell to have to add and layout the views each time it's recycled which would hurt performance. Is it possible to instead keep the `MenuViewItem` more static with the views remaining but only have the text/image contents updated? If there is a need for a different layout, it would be better to create another subclass to handle that.

sleroux replied:
I guess the above only applies if you're recycling so it doesn't matter :/

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+            if recogniser is UILongPressGestureRecognizer {
+                self.removeGestureRecognizer(recogniser)
+            }
+        }
```
Nit: Can clean this up a bit:

```swift
gestureRecognizers?.filter { $0 is UILongPressGestureRecognizer}. forEach { self.removeGestureRecognizer($0) }
```

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,106 @@
+        titleLabel.text = title
+    }
+
+    public func setImage(image: UIImage) {
```
If you have these accessor methods for `imageView` and `textLabel` you probably don't need to expose those views publicly at all and keep them private.

sleroux commented on Client/Frontend/Menu/MenuView.swift:
```diff
@@ -0,0 +1,387 @@
+        super.init(frame: CGRectZero)
+
+
+        let pageControllerView = menuItemPageController.view
```
If you're adding the `menuItemPageController`'s view as a subview, I think you also should be following the UIViewControllerContainer steps to adding the view controller as a child as well [1]

[1] https://developer.apple.com/library/ios/featuredarticles/ViewControllerPGforiPhoneOS/ImplementingaContainerViewController.html

sleroux commented:
I notice that a lot of the code involves the paging view controller, laying out the menu view items with the toolbar, and implementing the data source/delegates for the menu view. I think you could make your life easier by using a UICollectionView and implementing `targetContentOffsetForProposedContentOffset` to handle the 'paging'. Some advantages:

1. Cell recycling in the case of switching pages.
2. A 'layout' abstraction that you can implement if you choose to layout both the toolbar and items in a collection view. Alternatively you can stick with the default flow layout for the items and layout the toolbar separately. I think this would clean up a lot of the code in `layoutSubviews` you have to calculate heights and such.
3. Leverage the existing data source/delegate methods UICollectionView gives you.

fluffyemily commented:
I hadn't thought about using a collection view for the page controller. That makes total sense. Thanks :)

fluffyemily commented:
@sleroux I've updated the paging code to use a CollectionVC and cleaned up the layout/init code for MenuView and MenuItemView. Wanna take another look?

sleroux commented on Client.xcodeproj/project.xcworkspace/xcshareddata/Client.xcscmblueprint:
```diff
@@ -23,6 +25,11 @@
+    },
+    {
       "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:mozilla\/firefox-ios.git",
       "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
```
Not sure where these changes came from

fluffyemily replied:
that'll probably be through a rebase against master that I didn't push from
the parent branch. Let me fix that.

On 23 March 2016 at 18:04, Stephan Leroux <notifications@github.com> wrote:

> In Client.xcodeproj/project.xcworkspace/xcshareddata/Client.xcscmblueprint
> <https://github.com/mozilla/firefox-ios/pull/1620#discussion_r57207949>:
>
> > @@ -23,6 +25,11 @@
> >    "DVTSourceControlWorkspaceBlueprintRelativePathToProjectKey" : "Client.xcodeproj",
> >    "DVTSourceControlWorkspaceBlueprintRemoteRepositoriesKey" : [
> >      {
> > +      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "https:\/\/github.com\/fluffyemily\/firefox-menu-experiments.git",
> > +      "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
> > +      "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey" : "17D28135537A0534CA3953E850B2215BF7F6515F"
> > +    },
> > +    {
> >        "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey" : "github.com:mozilla\/firefox-ios.git",
> >        "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey" : "com.apple.dt.Xcode.sourcecontrol.Git",
>
> Not sure where these changes came from
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/firefox-ios/pull/1620/files/d472ac87098126fd4eebbcc56e73b5b91eeb762d#r57207949>
>

sleroux commented on Client/Frontend/Menu/ItemPages/MenuPageCollectionViewCell.swift:
```diff
@@ -0,0 +1,28 @@
+    var pageView: MenuPageView = MenuPageView()
+
+    override init(frame: CGRect) {
+        super.init(frame: CGRectZero)
```
Nit: You can just pass in the frame thats passed in above. Not sure what happens if you just set this to CGRectZero.

fluffyemily replied:
missed this whenconverting a `UIView` into a `CollectionViewCell`. Doh!

sleroux commented on Client/Frontend/Menu/ItemPages/MenuPageCollectionViewCell.swift:
```diff
@@ -0,0 +1,28 @@
+//
```
MPL

fluffyemily replied:
*every time* :disappointed:

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -0,0 +1,54 @@
+        return menuTitleLabel
+    }()
+
+    private var previousLocation: CGPoint?
```
Nit: Don't think you need this line anymore

sleroux commented:
I'm a bit confused as to what purpose the `MenuPageView` has now that the MenuView has a collectionView containing the items. How does the `MenuPageView` work?

fluffyemily commented:
@sleroux each cell of the collection view is a menu page. Each menu page contains a selection of menu items. We scroll between the pages. 

To take a look at the case I am trying to solve with this design, change 
```
let mvc = MenuViewController(withMenuConfig: MenuConfiguration.menuConfigurationForLocation(.TabTray), presentationStyle: .Modal)
```
to
```
let mvc = MenuViewController(withMenuConfig: MenuConfiguration.menuConfigurationForLocation(.Browser), presentationStyle: .Modal)
```
to see a case where we more than one page of menu items to display.

When I tried to just add the menu items as a collection view, that pages horizontally, I ended up with 1 very long row of cell, with no wrapping within cells, or multiple rows that filled the top row first, and then the bottom, rather than the first page then the second. 

I then looked at having a collection view inside a collection view, with the internal collection view laying out the menu cells and the outer one dealing with the paging, but that became far too complicated very quicky.

I then explored creating a custom layout that did the correct wrapping, but again the solution I left with here felt much simpler to understand overall.

If there is a simple way of getting a multi row, horizontally scrolling collection view, that lays out in the desired manner, that I missed, please let me know.

jhugman commented on Client/Frontend/Browser/BrowserViewController.swift:
```diff
@@ -1378,7 +1378,7 @@ extension BrowserViewController: BrowserToolbarDelegate {
                 } else {
-                    self.addBookmark(url, title: tab.title)
+                    self.addBookmark(url, title: self.title)
                 }
```
Nit: Comment needed? As the bookmark is about the tab and the title, unsure why you made this change.

fluffyemily replied:
I made this change because things didn't compile without it, but I think I committed it accidentally.

jhugman commented on Client/Frontend/Menu/MenuConfiguration.swift:
```diff
@@ -0,0 +1,162 @@
+        case .TabTray:
+           return 4
+        default:
+            return 3
```
Nit: Whitespace.

jhugman commented on Client/Frontend/Menu/MenuViewController.swift:
```diff
@@ -0,0 +1,194 @@
+}
+extension MenuViewController: MenuItemDelegate {
+    func menuView(menu: MenuView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
+    }
```
Aha!

jhugman commented on Client/Frontend/Menu/MenuViewController.swift:
```diff
@@ -0,0 +1,194 @@
+
+extension MenuViewController: MenuToolbarDataSource {
+    func numberOfToolbarItemsInMenuView(menuView: MenuView) -> Int {
+        guard let menuToolbarItems = menuConfig.menuToolbarItems else { return 0}
```
Nit: Whitespace after `0`. `{ return 0}`

jhugman commented on Client/Frontend/Menu/MenuViewController.swift:
```diff
@@ -0,0 +1,194 @@
+}
+
+extension NSIndexPath {
+    func getMenuItemIndex() -> Int {
```
This is adding an instance method to every instance of `NSIndexPath`, even those not involved in the menu?

This feels… weird, especially if we're considering publishing the menu as a separate component.

fluffyemily replied:
I agree. In a later branch (to be PR'd soon) this will go away.

jhugman commented on Client/Frontend/Menu/Toolbar/MenuToolbarItem.swift:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    func iconForMode(isPrivate isPrivate: Bool = false) -> UIImage?  {
+        return isPrivate ? privateModeIcon : icon
```
For all these `isPrivate` trinaries: something doesn't feel quite right about it.

Are there any other modes we may feasibly support? e.g. Guest Mode.

fluffyemily replied:
These are temporary until we have created the `AppState`, at which time they will be refined to be less cumbersome.

jhugman commented:
I really like the way this is going.

I'm looking forward to seeing actions added. Good work!

sleroux commented on Client/Frontend/Menu/Items/PagingMenuItemCollectionViewLayout.swift:
```diff
@@ -0,0 +1,76 @@
+    var maxNumberOfItemsPerPageRow: Int = 0
+    var lineSpacing: CGFloat = 0
+    var interitemSpacing: CGFloat = 0
+    var cellSize: CGSize = CGSize.zero
```
I had no idea CGSize had a size property! And all this time I've been using CGSizeZero...

sleroux commented on Client/Frontend/Menu/Items/PagingMenuItemCollectionViewLayout.swift:
```diff
@@ -0,0 +1,76 @@
+    override func prepareLayout() {
+        guard let collectionView = self.collectionView,
+            let collectionViewDataSource = collectionView.dataSource else {
+                return super.prepareLayout()
```
Nit: Not sure if it matters but usually you should always call super. Would it make sense to move this above and do a plain return here?

sleroux commented on Client/Frontend/Menu/MenuView.swift:
```diff
@@ -298,7 +299,19 @@ class MenuView: UIView {
+        // calculate cell size
+        var cellWidth = (self.bounds.size.width - itemPadding) / numberOfItemsInRow
+        cellWidth -= itemPadding
+        let cellHeight = CGFloat(menuRowHeight)
```
I wonder if it's possible to move these calculations into the layout? Instead of using self.bounds, you'd use the connected collection view's bounds. I think you can also infer the height info from that too.

sleroux commented on Client/Frontend/Menu/MenuView.swift:
```diff
@@ -335,67 +348,38 @@ class MenuView: UIView {
-
-                itemsForPageIndex.append(view)
+        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(pagingCellReuseIdentifier, forIndexPath: indexPath) as! MenuItemCollectionViewCell
+        if let cellView = menuItemDataSource?.menuView(self, viewForItemAtIndexPath: indexPath) {
```
Can we merge the cellView with the MenuItemCollectionViewCell now that we have the single collection view or do we use the view elsewhere in a different context?

sleroux commented on Client/Frontend/Menu/Items/MenuItemView.swift:
```diff
@@ -44,16 +37,18 @@ public class MenuItemView: UIView {
 
-    func prepareForReuse() {
+    override func prepareForReuse() {
         menuImageView.image = nil
```
Nit: Might want to call super.prepareForReuse here

sleroux commented on Client/Frontend/Menu/Items/PagingMenuItemCollectionViewLayout.swift:
```diff
@@ -7,9 +7,10 @@ import UIKit
     var lineSpacing: CGFloat = 0
     var interitemSpacing: CGFloat = 0
-    var cellSize: CGSize = CGSize.zero
+//    var cellSize: CGSize = CGSize.zero
```
Nit: Comment
